src/device_database.rs

```diff
@@ -0,0 +1,627 @@
+            room: room.map(|s| s.to_string()),
+            discovered_via: source,
+            first_seen: now,
+            last_seen: now,
+            last_api_sync: Some(now),
+            user_name: None,
+            user_room: None,
+        }
+    }
+}
+
+/// Compute a device name from SKU and ID (matches Device::computed_name logic)
+fn compute_device_name(sku: &str, id: &str) -> String {
+    // The id is usually "XX:XX:XX:XX:XX:XX:XX:XX" but some devices
+    // report it without colons, and in lowercase. Normalize it.
+    let mut normalized_id = String::new();
+    for c in id.chars() {
+        if c == ':' {
+            continue;
+        }
+        normalized_id.push(c.to_ascii_uppercase());
+    }
+
+    format!(
+        "{}_{}",
+        sku,
+        &normalized_id[normalized_id.len().saturating_sub(4)..]
+    )
+}
+
+impl DeviceDatabase {
+    /// Load database from disk, or create empty if not exists
+    pub fn load(path: &Path) -> anyhow::Result<Self> {
+        if path.exists() {
+            let contents = fs::read_to_string(path)?;
+            let db: DeviceDatabase = serde_json::from_str(&contents)?;
+            log::info!(
+                "Loaded device database with {} devices from {:?}",
+                db.devices.len(),
+                path
+            );
+            Ok(db)
+        } else {
+            log::info!(
+                "No device database found at {:?}, starting fresh",
+                path
+            );
+            Ok(DeviceDatabase::default())
+        }
+    }
+
+    /// Save database atomically (write temp file, then rename)
+    pub fn save(&self, path: &Path) -> anyhow::Result<()> {
+        // Create parent directory if needed
+        if let Some(parent) = path.parent() {
+            fs::create_dir_all(parent)?;
+        }
+
+        // Write to temporary file in same directory
+        let temp_path = path.with_extension("json.tmp");
+        let contents = serde_json::to_string_pretty(self)?;
+
+        {
+            let mut file = fs::File::create(&temp_path)?;
+            file.write_all(contents.as_bytes())?;
+            file.sync_all()?; // Ensure data is on disk
+        }
+
+        // Atomic rename
+        fs::rename(&temp_path, path)?;
+
+        log::debug!(
+            "Saved device database with {} devices to {:?}",
+            self.devices.len(),
+            path
+        );
+
+        Ok(())
+    }
+
+    /// Get the default database path based on environment
+    pub fn default_path() -> PathBuf {
+        // Check for explicit configuration first
+        if let Ok(path) = std::env::var("GOVEE_DEVICE_DB") {
+            return PathBuf::from(path);
+        }
+
+        // Home Assistant add-on uses /data
+        let addon_path = PathBuf::from("/data/devices.json");
+        if addon_path
+            .parent()
+            .map(|p| p.exists())
+            .unwrap_or(false)
+        {
+            return addon_path;
+        }

```

Comment:
This could be a bit funky because it will use `/data/devices.json` if run on any system that happens to have `/data` on the filesystem.

I think it would be best overall (but sadly more fussy in the code) to add an optional command line argument for this to `ServeCommand` (clap's `arg` stuff should make this easy) and then use `opt_env_var` to facilitate the env var configuration mode.

Then in `addon/run.sh` we can set the command line argument or env var explicitly for the addon.